src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -243,6 +245,11 @@ public override List<TSource> ToList()
                 return list;
             }
 
+            public override HashSet<TSource> ToHashSet(IEqualityComparer<TSource> comparer)
+            {
+                return new HashSet<TSource>(_source, comparer) { _item };
```

Comment:
If we change this to `_source.ToHashSet(comparer)` then if `_source` is itself a type with an optimised path from this PR it will be able to take that path.